File: ios/Sources/AppAttestClient/RequestHash.swift
import CryptoKit
import Foundation

/// 把一個 HTTP 請求綁定到一個 challenge 上。
///
/// ┌─────────────────────────────────────────────────────────────────┐
/// │ 這段必須跟後端的 src/RequestHash.php 逐位元組完全一致。          │
/// │ 改這裡就等於改合約，兩邊必須同時改、同時發版。                  │
/// └─────────────────────────────────────────────────────────────────┘
///
/// 算法：
///
///     requestHash = base64url_nopad(
///         SHA256( METHOD ‖ "\n" ‖ path ‖ "\n" ‖ SHA256(body) ‖ "\n" ‖ challenge )
///     )
///
/// - `SHA256(body)` 是 **32 bytes 的原始位元組**，不是 hex 字串
/// - `challenge` 是 server 給的**原始 32 bytes**，不是 base64 字串
/// - `path` 不含 query string，而且要用 **percent-encoded** 的形式（見下）
///
/// `RequestHashTests` 裡的黃金測試向量跟 PHP 端 `RequestHashTest` 用的是
/// 同一組值。兩邊任一方改動導致向量對不上，測試就會紅。
public enum RequestHash {

    /// 算出一個請求的 requestHash。
    ///
    /// - Parameters:
    ///   - method: HTTP method，大小寫不拘（內部會轉大寫）
    ///   - path: URL path，不含 query string，percent-encoded
    ///   - body: 原始 request body（沒有 body 就傳 `Data()`）
    ///   - challenge: server 發的原始 32 bytes
    public static func compute(method: String, path: String, body: Data, challenge: Data) -> String {
        var payload = Data()

        payload.append(Data(method.uppercased().utf8))
        payload.append(0x0A)                                 // "\n"
        payload.append(Data(path.utf8))
        payload.append(0x0A)
        payload.append(Data(SHA256.hash(data: body)))        // 原始 32 bytes，不是 hex
        payload.append(0x0A)
        payload.append(challenge)                            // 原始 32 bytes，不是 base64

        return Data(SHA256.hash(data: payload)).base64URLEncodedString()
    }

    /// 從 `URLRequest` 取出算 requestHash 需要的 path。
    ///
    /// === 這裡有一個很容易踩的跨語言陷阱 ===
    ///
    /// `URL.path` 回傳的是 **percent-decoding 之後**的字串：
    ///
    ///     URL(string: "https://x.com/a%20b")!.path        // "/a b"    ← 解碼過
    ///     components.percentEncodedPath                    // "/a%20b"  ← 線上實際傳的
    ///
    /// 後端看到的是 `REQUEST_URI`，也就是**線上實際傳的那份**（未解碼）。
    /// 所以這裡必須用 `percentEncodedPath`，否則只要 path 含有非 ASCII 字元
    /// 或 `%XX`，兩邊算出的 hash 就會不同 —— 而且只有那些特定的 URL 會壞，
    /// 其他一切正常，是最難查的那種 bug。
    ///
    /// 沒有 path 時回傳 `"/"`，與後端行為一致。
    public static func path(of url: URL) -> String {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return "/"
        }

        let path = components.percentEncodedPath

        return path.isEmpty ? "/" : path
    }
}

extension Data {
    /// base64url（RFC 4648 §5），無 padding。
    ///
    /// 跟一般 base64 的差別只有：`+` → `-`、`/` → `_`，並去掉結尾的 `=`。
    /// 這樣才能安全地放進 HTTP header 和 URL。
    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    /// 解 base64url。輸入不合法時回傳 `nil`。
    static func fromBase64URL(_ string: String) -> Data? {
        var base64 = string
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")

        // 補回 base64 需要的 padding。
        let remainder = base64.count % 4
        if remainder > 0 {
            base64 += String(repeating: "=", count: 4 - remainder)
        }

        return Data(base64Encoded: base64)
    }
}

File: src/RequestHash.php
<?php

declare(strict_types=1);

namespace AppAttest;

/**
 * 把一個 HTTP 請求綁定到一個 challenge 上。
 *
 * ┌─────────────────────────────────────────────────────────────────┐
 * │ 這是整個系統裡唯一一段「iOS 端必須逐位元組寫成一模一樣」的邏輯。 │
 * │ 改這裡就等於改 client 合約，兩邊必須同時改、同時發版。           │
 * └─────────────────────────────────────────────────────────────────┘
 *
 * 算法：
 *
 *   requestHash = base64url_nopad(
 *       SHA256( method ‖ "\n" ‖ path ‖ "\n" ‖ SHA256(body) ‖ "\n" ‖ challenge )
 *   )
 *
 * 其中：
 *   method    大寫 ASCII，例如 "POST"
 *   path      不含 query string 的 URL path，例如 "/protected/echo"
 *   SHA256(body)  原始 body 位元組的 **raw binary** 摘要（32 bytes，不是 hex）
 *   challenge     challengeId 對應的 **原始 32 bytes**（不是 base64 字串）
 *   ‖         位元組串接
 *
 * 為什麼要把 challenge 混進去：這樣 requestHash 天生就綁死這一次的
 * challenge。攻擊者攔到一組合法的 (requestHash, assertion) 之後，沒辦法
 * 拿去配另一個 challenge 用 —— 而 challenge 是一次性的，所以整組憑證
 * 只能用一次。
 *
 * 為什麼要把 method 和 path 混進去：否則攔截者可以把一個打到
 * /protected/echo 的合法簽章，原封不動轉送到 /admin/delete-everything。
 * body 相同時簽章一樣會過。
 */
final class RequestHash
{
    /**
     * 算出一個請求的 requestHash。
     *
     * @param string $method    HTTP method，大寫
     * @param string $path      URL path，不含 query string，且必須是**未解碼**的形式
     * @param string $body      原始 request body（空 body 傳 ""）
     * @param string $challenge challenge 的原始 32 bytes
     *
     * === $path 一定要用未解碼的形式 ===
     *
     * client 送出的是 percent-encoded 的 path（線上實際傳的那份），所以
     * server 也必須用同一份去算。
     *
     *     parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH)   // ✅ 未解碼
     *     urldecode(...)                                      // ❌ 解碼過就對不上
     *
     * 很多框架的 router 會貼心地幫你把 path 解碼後才交給 controller。若你
     * 們的框架是這樣，這裡要改成讀原始的 REQUEST_URI，否則只有含 %XX 或
     * 非 ASCII 的 URL 會驗不過，其他一切正常 —— 極難查。
     */
    public static function compute(string $method, string $path, string $body, string $challenge): string
    {
        $payload = strtoupper($method)
            . "\n"
            . $path
            . "\n"
            // 注意第三個參數 true：要的是 32 bytes 的 raw binary。
            // 漏掉就會變成 64 字元的 hex 字串，算出來的 hash 完全不同，
            // 而且症狀是「簽章明明對卻一直驗不過」，極難查。
            . hash('sha256', $body, true)
            . "\n"
            . $challenge;

        return Base64Url::encode(hash('sha256', $payload, true));
    }

    /**
     * 常數時間比對 client 送來的 requestHash 與 server 重算的結果。
     *
     * 用 hash_equals() 而不是 ===，避免比對耗時洩漏「前幾個字元對了」
     * 這種資訊。這裡的 timing 攻擊實際上很難打（值每次都不同），但這是
     * 一行就能做到的事，沒有理由不做。
     */
    public static function matches(string $fromClient, string $recomputed): bool
    {
        return hash_equals($recomputed, $fromClient);
    }
}
